File: python/1035. Uncrossed Lines.py
class Solution(object):
    def maxUncrossedLines(self, A, B):
        """
        :type A: List[int]
        :type B: List[int]
        :rtype: int
        """
        la, lb = len(A), len(B)
        dp = [[0]*(lb+1) for _ in range(la+1)]
        for i in range(1, la+1):
            for j in range(1, lb+1):
                val = int(A[i-1]==B[j-1])
                dp[i][j] = max(dp[i-1][j-1]+val, dp[i-1][j], dp[i][j-1])
        return dp[la][lb]
        

if __name__ == '__main__':
    A = [2,5,1,2,5]
    B = [10,5,2,1,5,2]
    su = Solution()
    su.maxUncrossedLines(A, B)

File: go/src/典型问题/DP/1035. Uncrossed Lines/1035. Uncrossed Lines.go
package main

import "math"

func maxUncrossedLines(nums1 []int, nums2 []int) int {
	n1, n2 := len(nums1), len(nums2)
	dp := make([][]int, n1+1)
	for i := 0; i <= n1; i++ {
		dp[i] = make([]int, n2+1)
	}
	for i := 1; i <= n1; i++ {
		for j := 1; j <= n2; j++ {
			t := 0
			if nums1[i-1] == nums2[j-1] {
				t = 1
			}
			dp[i][j] = max(dp[i-1][j-1] + t, dp[i-1][j], dp[i][j-1])
		}
	}
	return dp[n1][n2]
}

func max(a ...int) int {
	var maxVal = math.MinInt32
	for _, i := range a {
		if i > maxVal {
			maxVal = i
		}
	}
	return maxVal
}
